src/fmu/ensemble/realization.py

```diff
@@ -675,25 +675,61 @@ def get_df(self, localpath):
           * simulator_volume_fipnum.csv
           * share/results/volumes/simulator_volume_fipnum
 
-        but only as long as there is no ambiguity. In case of ambiguity, a
-        ValueError will be raised.
+        but only as long as there is no ambiguity. In case of ambiguity,
+        you will get localpath as is if it exists, if not, a KeyError.
 
         Args:
             localpath (str): the idenfier of the data requested
+            merge (list or str): identifier/localpath of some data to be merged in,
+                typically 'parameters.txt'. Will only work when return type is a
+                dataframe. If list is supplied, order can matter.
 
         Returns:
-            dataframe or dictionary
+            dataframe or dictionary.
 
         Raises:
-            ValueError if data is not found.
+            KeyError if data is not found.
         """
-        if localpath in self.data.keys():
-            return self.data[localpath]
         fullpath = self.shortcut2path(localpath)
-        if fullpath in self.data.keys():
-            return self.data[self.shortcut2path(localpath)]
-        # KeyError would also be valid or better here.
-        raise ValueError("Could not find {}".format(localpath))
+        if fullpath not in self.data.keys():
+            raise KeyError("Could not find {}".format(localpath))
+        data = self.data[self.shortcut2path(localpath)]
+        if not isinstance(merge, list):
+            merge = [merge]  # can still be None
+        if merge and merge[0] is not None:
+            # Strange things can happen when we do merges since
+            # this function happily returns references to the internal
+            # dataframes in the realization object. So ensure
+            # we copy dataframes if any merging is about to happen.
+            if isinstance(data, pd.DataFrame):
+                data = data.copy()
+            elif isinstance(data, dict):
+                data = data.copy()
```

Comment:
Dicts with mutable types would break this shallow copy, but such dicts are not intentionally supported, and not possible to load unless they are injected into the objects by a hack. So the fix would be to elsewhere test dict objects for types, or just let the hackers deal with their own errors..